src/core/lib/transport/metadata.h

```diff
@@ -132,17 +134,111 @@ void* grpc_mdelem_get_user_data(grpc_mdelem md, void (*if_destroy_func)(void*));
 void* grpc_mdelem_set_user_data(grpc_mdelem md, void (*destroy_func)(void*),
                                 void* user_data);
 
+typedef void (*destroy_user_data_func)(void* user_data);
+
+struct UserData {
+  gpr_mu mu_user_data;
+  gpr_atm destroy_user_data;
+  gpr_atm user_data;
+};
+
+struct interned_metadata {
+  /* must be byte compatible with grpc_mdelem_data */
+  grpc_slice key;
+  grpc_slice value;
+
+  /* private only data */
+  gpr_atm refcnt;
```

Comment:
I'd suggest converting this class to C++, adding a Ctor and using grpc_core::RefCount here and else where for refcount. It's probably another 10 lines of change but you'll get fully inlined ref counts.

Reply:
Done.